### DISCUSSION TITLE
afterStateUpdated() in Edit and View pages not return records

### DISCUSSION THREAD

**Discussion by Professor93M** [filamentphp/filament#9014] [Help] [answered]
Dears,

I have this code:

```
Forms\Components\Select::make('invoices_id')
    ->label(__('filament-panels::SheetsResource/forms.invoices_id'))
    ->relationship('invoices', 'id')
    ->reactive()
    ->live()
    ->afterStateUpdated(function ($state, Set $set) {
        $invoice = Invoices::find($state);
        $customers_id = $invoice->customers()->get()[0]->fullname;
        $set('customers_id', $customers_id);
    })
    ->required(),
Forms\Components\TextInput::make('customers_id')
    ->label(__('filament-panels::CustomersResource/forms.fullname'))->disabled(),
```

It's working well, but the problem is when go to the edit page of the view page it does not return **customers_id** and it's still empty
is there any way to fill this field in these pages or use another method instead of afterStateUpdated, or maybe I'm doing something wrong?

**Comment by joshembling** [comment, marked as answer]
You will probably want to use `afterStateHydrated()` for this as you want to fill the state, am I right? 

Something like:
```php
->afterStateHydrated(function (Select $component, ?string $state, Get $get) {
    $invoice = Invoices::find($state);
    $customers_id = $invoice->customers()->get()[0]->fullname;

    $component->state($customers_id);
})
```

**Reply by Professor93M** [reply]
thanks for your response, actually I used your method it worked with the Edit and View but also needs afterStateUpdated() at the Create page
so I modified the code like that and it's ok now:

```
->afterStateUpdated(function ($state, Set $set) {
        $invoice = Invoices::find($state);
        $customers_id = $invoice->customers()->get()[0]->fullname;
        $set('customers_id', $customers_id);
    })->afterStateHydrated(function (Select $component, ?string $state, Get $get) {
    $invoice = Invoices::find($state);
    $customers_id = $invoice->customers()->get()[0]->fullname;

    $component->state($customers_id);
})
```
the two methods work together **afterStateUpdated** for Create Page and **afterStateHydrated** for Edit and View.
now it works well but is it ok for now like this code or need any modification?

thank you.

**Reply by joshembling** [reply]
Yes it will not work on create because the database record doesn’t exist, therefore can’t hydrate the model. 

Glad that’s working for you 🎉
